File: roms/super_real_mahjong_part3.py
#
#	Super Real Mahjong Part 3
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    merge = lambda even, odd : bytes([odd[i // 2] if i % 2 else even[i // 2] for i in range(len(even) * 2)])
    prg = z.read('za0-10.bin')
    gfx = merge(z.read('za0-02.bin') + z.read('za0-01.bin') + z.read('za0-06.bin') + z.read('za0-05.bin'), z.read('za0-04.bin') + z.read('za0-03.bin') + z.read('za0-08.bin') + z.read('za0-07.bin'))
    voi = z.read('za0-11.bin')
    color_h = z.read('za0-12.prm')
    color_l = z.read('za0-13.prm')

rom = prg + gfx + voi + color_h + color_l

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/super_real_mahjong_part3.js
/*
 *	Super Real Mahjong Part 3
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};
const merge = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0)), step = args.length;
	return args.forEach((a, i) => a.forEach((e, j) => array[j * step + i] = e)), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg = z.read('za0-10.bin');
const gfx = merge(cat(z.read('za0-02.bin'), z.read('za0-01.bin'), z.read('za0-06.bin'), z.read('za0-05.bin')), cat(z.read('za0-04.bin'), z.read('za0-03.bin'), z.read('za0-08.bin'), z.read('za0-07.bin')));
const voi = z.read('za0-11.bin');
const color_h = z.read('za0-12.prm');
const color_l = z.read('za0-13.prm');

const rom = cat(prg, gfx, voi, color_h, color_l);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
